lib/ansible/modules/cloud/smartos/vmadm.py

```diff
@@ -0,0 +1,600 @@
+        description:
+          - Default value for a virtual NIC model for KVM guests.
+    nics:
+        required: false
+        description:
+          - A list of nics to add, valid properties are documented in vmadm(1M).
+    qemu_extra_opts:
+        required: false
+        description:
+          - Additional qemu cmdline arguments for KVM guests.
+    quota:
+        required: false
+        description:
+          - Quota on zone filesystems (in MiBs).
+    ram:
+        required: false
+        description:
+          - Amount of virtual RAM for a KVM guest (in MiBs).
+    resolvers:
+        required: false
+        description:
+          - List of resolvers to be put into C(/etc/resolv.conf).
+    uuid:
+        required: false
+        description:
+          - UUID of the VM. Can either be a full UUID or C(*) for all VMs.
+    vcpus:
+        required: false
+        description:
+          - Number of virtual CPUs for a KVM guest.
+    vnc_port:
+        required: false
+        description:
+        - TCP port to listen of the VNC server. Or set C(0) for random,
+          or C(-1) to disable.
+    zfs_io_priority:
+        required: false
+        description:
+          - IO throttle priority value relative to other VMs.
+    zpool:
+        required: false
+        description:
+          - ZFS pool the VM's zone dataset will be created in.
+requirements:
+    - python >= 2.6
+'''
+
+EXAMPLES = '''
+- name: create SmartOS zone
+  vmadm:
+    brand: joyent
+    state: present
+    alias: fw_zone
+    image_uuid: 95f265b8-96b2-11e6-9597-972f3af4b6d5
+    firewall_enabled: yes
+    indestructible_zoneroot: yes
+    nics:
+      - nic_tag: admin
+        ip: dhcp
+        primary: true
+    internal_metadata:
+      root_pw: 'secret'
+    quota: 1
+
+- name: Delete a zone
+  vmadm:
+    alias: test_zone
+    state: deleted
+
+- name: Stop all zones
+  vmadm:
+    uuid: '*'
+    state: stopped
+'''
+
+RETURN = '''
+uuid:
+    description: UUID of the managed VM.
+    returned: always
+    type: string
+    sample: 'b217ab0b-cf57-efd8-cd85-958d0b80be33'
+alias:
+    description: Alias of the managed VM.
+    returned: When addressing a VM by alias.
+    type: string
+    sample: 'dns-zone'
+state:
+    description: State of the target, after execution.
+    returned: success
+    type: string
+    sample: 'running'
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+import os
+import re
+import tempfile
+try:
+    import json
+except ImportError:
+    import simplejson as json
+
+# While vmadm(1M) supports a -E option to return any errors in JSON, the
+# generated JSON does not play well with the JSON parsers of Python.
+# The returned message contains '\n' as part of the stacktrace,
+# which breaks the parsers.
+
+
+def get_vm_prop(module, uuid, prop):
+    """Lookup a property for the given VM.
+    Returns the property, or None if not found."""
+    cmd = 'vmadm lookup -j -o {0} uuid={1}'.format(prop, uuid)
```

Comment:
vmadm is located on a read-only filesystem that cannot be moved to/from another location than `/usr/sbin/vmadm`. I will actually make `VMADM` globally available as `/usr/sbin/vmadm` since it's fixed and there is no other location it can reside.